Section Selection functionality with better conflict checking in place

meganli0818 opened the pull request:
This pull request includes significant changes to the `load_courses.py`, `style.css`, and `courses.js` files to enhance course handling, improve UI for section selection, and manage cart functionality. The most important changes include introducing section selection for courses, indicators for conflicting sections as well as sections in the cart, and updating cart management to handle sections and conflicts more effectively.

Enhancements to course handling:

* [`amherst_coursework_backend/amherst_coursework_algo/management/commands/load_courses.py`](diffhunk://#diff-1e95b539d208de84dc6a9bda6a5becfc782125406ddaaf351f4dcb495c306715R448-R464): Added logic to create a dummy section with a dummy professor for courses that have no sections.

UI improvements for section selection:

* [`amherst_coursework_backend/amherst_coursework_algo/static/css/style.css`](diffhunk://#diff-4de82f69dd12c7c534c5be93736646e435d778239f6cccf129c81ab3bd954877R619-R693): Added CSS for modal UI components, section items, and conflict badges to improve the user interface for selecting sections.

Cart functionality and conflict management:

* [`amherst_coursework_backend/amherst_coursework_algo/static/js/courses.js`](diffhunk://#diff-a012d80ac3fb72bc2f44165b60ea3effc2b08f92f49aec4c8f67a0849bb3c2e7L19-R44): Updated `handleCartClick` to manage sections within the cart, including adding/removing sections and updating button states.
* [`amherst_coursework_backend/amherst_coursework_algo/static/js/courses.js`](diffhunk://#diff-a012d80ac3fb72bc2f44165b60ea3effc2b08f92f49aec4c8f67a0849bb3c2e7L32-R171): Refactored `findAndMarkAllCartConflicts` to use `fetch` for conflict checking and apply visual conflict styling dynamically.
* [`amherst_coursework_backend/amherst_coursework_algo/static/js/courses.js`](diffhunk://#diff-a012d80ac3fb72bc2f44165b60ea3effc2b08f92f49aec4c8f67a0849bb3c2e7L536-R668): Added `showSectionModal` and `formatMeetingTimes` functions to display a modal for section selection and format meeting times for display.

Bug fixes:
* Fixed bug where some conflict badges improperly remained on course cards even after cart was cleared.

<!-- Korbit AI PR Description Start -->
## Description by Korbit AI

### What change is being made?
Integrate section selection functionality in the course cart management system while enhancing conflict checking and improving code organization across backend and frontend components.

### Why are these changes being made?
The previous implementation lacked detailed section-level granularity in course selection, leading to potential scheduling conflicts being overlooked. This update incorporates section-level awareness, enabling users to select specific sections of courses and allowing the system to check for and highlight time conflicts. This approach improves the user experience by providing more precise scheduling capabilities and preventing hidden conflicts within the cart. The changes also involve refactoring code for better readability, maintainability, and integration of institutional settings for consistency.

> Is this description stale? Ask me to generate a new description by commenting `/korbit-generate-pr-description`
<!-- Korbit AI PR Description End -->

meganli0818 committed:
Added section selection for adding course to cart and all associated calender/time conflict changes

meganli0818 committed:
Added conflict detection for course sections and shows conflicts when selecting sections

meganli0818 committed:
Fixed bug where course conflcits were not being fully removed and added dummy secitons for courses without sections

meganli0818 committed:
Added visual indicators for sections in the cart

meganli0818 committed:
Refine cart handling to remove specific course sections

meganli0818 commented:
New images!
<img width="1439" alt="Screenshot 2025-03-21 at 10 21 21 PM" src="https://github.com/user-attachments/assets/bca83f09-faae-4990-b628-ac47f70828b2" />

<img width="1438" alt="Screenshot 2025-03-21 at 10 21 52 PM" src="https://github.com/user-attachments/assets/d622befd-7adf-4914-9f0f-3d365fc7baa5" />

<img width="1440" alt="Screenshot 2025-03-21 at 10 22 29 PM" src="https://github.com/user-attachments/assets/df158af9-6c7d-44e5-88a0-4b05584e3afb" />

DhyeyMavani2003 commented on amherst_coursework_backend/amherst_coursework_algo/views.py:
```diff
@@ -107,78 +107,96 @@ def format_meeting_times(time_slots):
+            }
+            cart_data.append(course_info)
+        except Course.DoesNotExist:
+            print(f"ERROR: Course with ID {course_id} not found")
```
@meganli0818 address this!

DhyeyMavani2003 commented on amherst_coursework_backend/amherst_coursework_algo/views.py:
```diff
@@ -107,78 +107,96 @@ def format_meeting_times(time_slots):
+                            if section.friday_end_time
+                            else None
+                        ),
+                    }
```
@meganli0818 resolve this!

DhyeyMavani2003 commented on amherst_coursework_backend/amherst_coursework_algo/static/js/courses.js:
```diff
@@ -16,9 +16,32 @@ function debounce(func, wait) {
+        });
+        updateButtonState(courseId, true);
+    } else {
+        cart = cart.filter(item => item.courseId !== courseId || item.sectionId !== sectionId);
+        containsCourse = cart.filter(item => item.courseId === courseId);
```
@meganli0818 resolve this!

DhyeyMavani2003 commented on amherst_coursework_backend/amherst_coursework_algo/static/js/courses.js:
```diff
@@ -29,183 +52,151 @@ function toggleCartPanel() {
-    
-    // If cart is empty, we're done (all styling cleared)
-    if (cartIds.length === 0) {
-        localStorage.removeItem('courseTimeConflicts'); // Clear stored conflicts
-        return;
-    }
-    
-    // Initialize courseConflicts object to store all conflicts
+    const cart = JSON.parse(localStorage.getItem('courseCart') || '[]');
     const courseConflicts = {};
     
-    // Process each cart course
-    cartIds.forEach(cartCourseId => {
-        // Find the course card
-        const cartCard = document.querySelector(`.course-card[data-course-id="${cartCourseId}"]`);
-        if (!cartCard) {
-            return; // Skip this iteration
-        }
-        
-        // Get the course name for better messages
-        const cartCourseName = cartCard.querySelector('.course-name')?.textContent || 'Unknown Course';
+    return new Promise(resolve => {
+        const cards = Array.from(document.querySelectorAll('.course-card'));
+        let completed = 0;
         
-        // Get the meeting time info
-        const cartInfoText = cartCard.querySelector('.info-text')?.textContent;
-        if (!cartInfoText) {
-            return;
-        }
+        cards.forEach(card => {
+            // Use requestAnimationFrame for smoother DOM updates
+            requestAnimationFrame(() => {
+                card.classList.remove('time-conflict');
+                card.style.border = '';
+                card.style.backgroundColor = '';
+                const existingBadge = card.querySelector('.conflict-badge');
+                if (existingBadge) {
+                    card.removeChild(existingBadge);
+                }
```
@meganli0818 resolve this if possible!

DhyeyMavani2003 commented on amherst_coursework_backend/amherst_coursework_algo/static/js/courses.js:
```diff
@@ -29,183 +52,151 @@ function toggleCartPanel() {
-    // Get stored conflicts
-    const courseConflicts = JSON.parse(localStorage.getItem('courseTimeConflicts') || '{}');
-    
-    // Check each course card for conflicts with cart courses
-    document.querySelectorAll('.course-card').forEach(card => {
-        const courseId = card.dataset.courseId;
-        
-        // Skip courses already in cart
-        if (cartIds.includes(courseId)) return;
-        
-        // Check if any cart course conflicts with this course
-        let hasConflict = false;
-        
-        cartIds.forEach(cartCourseId => {
-            // Check if this cart course has conflicts with the current card
-            if (courseConflicts[cartCourseId] && 
-                courseConflicts[cartCourseId].includes(courseId)) {
-                hasConflict = true;
-            }
+                    fetch(`/api/course/${courseId}/sections/`)
```
@meganli0818 use this API and resolve this comment from korbit-ai-bot

DhyeyMavani2003 commented on amherst_coursework_backend/amherst_coursework_algo/static/js/courses.js:
```diff
@@ -359,10 +350,10 @@ function updateCartDisplay() {
         // Create URLSearchParams object properly
         const params = new URLSearchParams();
-        cartIds.forEach(id => params.append('ids[]', id));
+        params.append('cart', JSON.stringify(cart));
```
@meganli0818 resolve this!

DhyeyMavani2003 commented on amherst_coursework_backend/amherst_coursework_algo/management/commands/load_courses.py:
```diff
@@ -445,6 +445,23 @@ def handle(self, *args, **options):
                     course.courseMaterialsLink = courseMaterialsLink
                     course.save()
 
+                    if course.sections.all().count() == 0:
+                        dummy_professor, _ = Professor.objects.get_or_create(
+                            name="TBA", link="https://www.amherst.edu/"
+                        )
+                        dummy_section, _ = Section.objects.update_or_create(
```
No need for this for now!

DhyeyMavani2003 commented on amherst_coursework_backend/amherst_coursework_algo/management/commands/load_courses.py:
```diff
@@ -445,6 +445,23 @@ def handle(self, *args, **options):
 
+                    if course.sections.all().count() == 0:
+                        dummy_professor, _ = Professor.objects.get_or_create(
+                            name="TBA", link="https://www.amherst.edu/"
```
@meganli0818 do this real quick! resolve this!

DhyeyMavani2003 reviewed (changes requested):
@meganli0818 great work on implementation! Please swiftly resolve the comments from korbit-ai for now (commented on ones that are needed before merge)! Let me know once you are done, and I can give it a final look!

meganli0818 committed:
Added suggestions from korbit, including the following:
- Minimizing api calls during course-cart conflict calculation
- Breaking down the course-cart conflict calculation into smaller functions
- Separating section time parsing into a separate function
- Using a global variable to store the institutional domain of amherst

DhyeyMavani2003 reviewed (approved):
LGTM!
